File: benchmarks/tsb/bench_dataframe_numeric_pipeline.ts
/**
 * Benchmark: DataFrame numeric pipeline — chain abs → round → sign on a 100k-row × 3-column DataFrame.
 * Tests a realistic sequence of standalone DataFrame numeric operations.
 * Outputs JSON: {"function": "dataframe_numeric_pipeline", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, dataFrameAbs, dataFrameRound, dataFrameSign } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 20;

const df = DataFrame.fromColumns({
  a: Array.from({ length: SIZE }, (_, i) => Math.sin(i * 0.01) * 150 - 20),
  b: Array.from({ length: SIZE }, (_, i) => Math.cos(i * 0.02) * 80),
  c: Array.from({ length: SIZE }, (_, i) => (i % 1000) * 0.123 - 50),
});

for (let i = 0; i < WARMUP; i++) {
  const a = dataFrameAbs(df);
  const b = dataFrameRound(a, { decimals: 1 });
  dataFrameSign(b);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  const a = dataFrameAbs(df);
  const b = dataFrameRound(a, { decimals: 1 });
  dataFrameSign(b);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dataframe_numeric_pipeline",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_dataframe_numeric_pipeline.py
"""
Benchmark: DataFrame numeric pipeline — chain abs → round → sign on a 100k-row × 3-column DataFrame.
Mirrors bench_dataframe_numeric_pipeline.ts.
Outputs JSON: {"function": "dataframe_numeric_pipeline", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import math
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 20

df = pd.DataFrame(
    {
        "a": [math.sin(i * 0.01) * 150 - 20 for i in range(SIZE)],
        "b": [math.cos(i * 0.02) * 80 for i in range(SIZE)],
        "c": [(i % 1000) * 0.123 - 50 for i in range(SIZE)],
    }
)

for _ in range(WARMUP):
    a = df.abs()
    b = a.round(1)
    np.sign(b)

start = time.perf_counter()
for _ in range(ITERATIONS):
    a = df.abs()
    b = a.round(1)
    np.sign(b)
total = (time.perf_counter() - start) * 1000

print(
    json.dumps(
        {
            "function": "dataframe_numeric_pipeline",
            "mean_ms": total / ITERATIONS,
            "iterations": ITERATIONS,
            "total_ms": total,
        }
    )
)
